### PR TITLE
Glibmm v2 toolchain

### PR DESCRIPTION
**Pull request by planetmarshall**
Specify library name and version:  **glibmm**

* Update glibmm for v2 toolchain
* Add version 2.74.0

---

- [x] I've read the [contributing guidelines](https://github.com/conan-io/conan-center-index/blob/master/CONTRIBUTING.md).
- [x] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
- [x] I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.
- [x] I've tried at least one configuration locally with the [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

File: recipes/glibmm/all/conanfile.py [hunk 1]
```diff
@@ -3,16 +3,26 @@
...
 from conan import ConanFile
-from conan.tools import (
-    build,
-    files,
-    microsoft,
-    scm
-)
 from conan.errors import ConanInvalidConfiguration
-from conans import Meson, tools
+from conan.tools.build import check_min_cppstd, cross_building
```
Comment (SpaceIm) [inline recipes/glibmm/all/conanfile.py:7]:
```suggestion
from conan.tools.apple import fix_apple_shared_install_name
from conan.tools.build import check_min_cppstd, cross_building
```

File: recipes/glibmm/all/conanfile.py [hunk 2]
```diff
@@ -176,7 +172,7 @@ def package(self):
                 )
 
         for dir_to_remove in ["pkgconfig", self._glibmm_lib, self._giomm_lib]:
-            files.rmdir(self, os.path.join(self.package_folder, "lib", dir_to_remove))
+            rmdir(self, os.path.join(self.package_folder, "lib", dir_to_remove))
```
Comment (SpaceIm) [inline recipes/glibmm/all/conanfile.py:175]:
```suggestion
            rmdir(self, os.path.join(self.package_folder, "lib", dir_to_remove))
            fix_apple_shared_install_name(self)
```

**Review by SpaceIm — CHANGES_REQUESTED** [review]

File: recipes/glibmm/all/conanfile.py [hunk 3]
```diff
@@ -43,129 +51,118 @@ def _giomm_lib(self):
...
         if self.settings.compiler.get_safe("cppstd"):
             if self._abi_version == "2.68":
-                build.check_min_cppstd(self, 17)
+                check_min_cppstd(self, 17)
             else:
-                build.check_min_cppstd(self, 11)
+                check_min_cppstd(self, 11)
+
+        if is_msvc(self) and not self.options.shared:
+            raise ConanInvalidConfiguration("Static library build is not supported by MSVC")
```
Comment (SpaceIm) [inline recipes/glibmm/all/conanfile.py:64]: It's a new limitation? Why did it work previously? Anyway it's a problem to raise for default option value.

File: recipes/glibmm/all/conanfile.py [hunk 4]
```diff
@@ -43,129 +51,118 @@ def _giomm_lib(self):
...
     def requirements(self):
-        self.requires("glib/2.75.0")
-
+        self.requires("glib/2.75.2")
         if self._abi_version == "2.68":
             self.requires("libsigcpp/3.0.7")
         else:
             self.requires("libsigcpp/2.10.8")
 
+    def generate(self):
```
Comment (SpaceIm) [inline recipes/glibmm/all/conanfile.py:92]: generate() after source() please, otherwise it's hard to understand flow of execution and global logic of the recipe.

File: recipes/glibmm/all/conanfile.py [hunk 5]
```diff
@@ -43,129 +51,118 @@ def _giomm_lib(self):
...
+        tc.generate()
+
+        env = VirtualBuildEnv(self)
+        env.generate()
+
     def source(self):
-        files.get(self, **self.conan_data["sources"][self.version], strip_root=True, destination=self._source_subfolder)
+        get(self, **self.conan_data["sources"][self.version], strip_root=True, destination=self.source_folder)
+
+    def export_sources(self):
```
Comment (SpaceIm) [inline recipes/glibmm/all/conanfile.py:111]: export_source() before config_option() please

File: recipes/glibmm/all/conanfile.py (hunk 3 above)
Comment (planetmarshall) [reply recipes/glibmm/all/conanfile.py]: MSVC static builds are disabled in the official sources, we've renabled them via a patch. I've reinstated support and added a patch for 2.75.0

File: recipes/glibmm/all/conanfile.py [hunk 6]
```diff
@@ -43,129 +51,115 @@ def _giomm_lib(self):
...
 
-        if self.options["glib"].shared and microsoft.is_msvc_static_runtime(self):
-            raise ConanInvalidConfiguration(
-                "Linking shared glib with the MSVC static runtime is not supported"
-            )
-
-    @property
-    def _source_subfolder(self):
-        return "source_subfolder"
+        if self.options["glib"].shared and is_msvc_static_runtime(self):
```
Comment (uilianries) [inline recipes/glibmm/all/conanfile.py:68]:
```suggestion
        if self.dependencies["glib"].options.shared and is_msvc_static_runtime(self):
```
self.dependencies is better than using old self.options on validate(): https://docs.conan.io/en/latest/reference/conanfile/dependencies.html

**Review by uilianries — COMMENTED** [review]
Only a small comment, otherwise, it's really good to go

**Review by uilianries — APPROVED** [review]
LGTM

File: recipes/glibmm/all/conanfile.py [hunk 7]
```diff
@@ -43,129 +51,115 @@ def _giomm_lib(self):
         return f"giomm-{self._abi_version}"
 
     def validate(self):
-        if hasattr(self, "settings_build") and build.cross_building(self):
+        if hasattr(self, "settings_build") and cross_building(self):
             raise ConanInvalidConfiguration("Cross-building not implemented")
```
Comment (SpaceIm) [inline recipes/glibmm/all/conanfile.py:55]: It should work now with MesonToolchain. It was a limitation of the legacy Meson build helper.

File: recipes/glibmm/all/conanfile.py [hunk 8]
```diff
@@ -43,129 +51,115 @@ def _giomm_lib(self):
...
 
         if self.settings.compiler.get_safe("cppstd"):
             if self._abi_version == "2.68":
-                build.check_min_cppstd(self, 17)
+                check_min_cppstd(self, 17)
             else:
-                build.check_min_cppstd(self, 11)
+                check_min_cppstd(self, 11)
 
         if self.options.shared and not self.options["glib"].shared:
```
Comment (SpaceIm) [inline recipes/glibmm/all/conanfile.py:63]:
```suggestion
        if self.options.shared and not self.dependencies["glib"].options.shared:
```

File: recipes/glibmm/all/conanfile.py [hunk 9]
```diff
@@ -43,129 +51,115 @@ def _giomm_lib(self):
...
         if self._abi_version == "2.68":
             self.requires("libsigcpp/3.0.7")
         else:
             self.requires("libsigcpp/2.10.8")
 
     def source(self):
-        files.get(self, **self.conan_data["sources"][self.version], strip_root=True, destination=self._source_subfolder)
+        get(self, **self.conan_data["sources"][self.version], strip_root=True, destination=self.source_folder)
+
+    def generate(self):
```
Comment (SpaceIm) [inline recipes/glibmm/all/conanfile.py:95]:
```suggestion
    def generate(self):
        env = VirtualBuildEnv(self)
        env.generate()
```

File: recipes/glibmm/all/conanfile.py [hunk 10]
```diff
@@ -43,129 +51,115 @@ def _giomm_lib(self):
...
+        tc.project_options.update({
+            "build-examples": "false",
+            "build-documentation": "false",
+            "msvc14x-parallel-installable": "false",
+            "default_library": "shared" if self.options.shared else "static"
+        })
+        tc.generate()
+
+        env = VirtualBuildEnv(self)
+        env.generate()
```
Comment (SpaceIm) [inline recipes/glibmm/all/conanfile.py:109]:
```suggestion
```

**Review by SpaceIm — CHANGES_REQUESTED** [review]
I strongly advise to always call VirtualBuildEnv first to avoid conflicts. VirtualBuildEnv is dumb, while other generators may sanitize similar env vars, so you don't want raw env vars of VirtualBuildEnv to override sanitized env vars of others clever generators.

File: recipes/glibmm/all/conanfile.py [hunk 11]
```diff
@@ -43,129 +51,115 @@ def _giomm_lib(self):
...
             del self.options.fPIC
 
+    def layout(self):
+        basic_layout(self, src_folder="src")
+
     def build_requirements(self):
-        self.build_requires("meson/0.64.1")
-        self.build_requires("pkgconf/1.9.3")
+        self.tool_requires("meson/1.0.0")
+        self.tool_requires("pkgconf/1.9.3")
```
Comment (SpaceIm) [inline recipes/glibmm/all/conanfile.py:83]:
```suggestion
        if not self.conf.get("tools.gnu:pkg_config", check_type=str):
            self.tool_requires("pkgconf/1.9.3")
```

File: recipes/glibmm/all/conanfile.py [hunk 12]
```diff
@@ -43,129 +51,115 @@ def _giomm_lib(self):
...
-
+        self.requires("glib/2.75.2")
         if self._abi_version == "2.68":
             self.requires("libsigcpp/3.0.7")
         else:
             self.requires("libsigcpp/2.10.8")
 
     def source(self):
-        files.get(self, **self.conan_data["sources"][self.version], strip_root=True, destination=self._source_subfolder)
+        get(self, **self.conan_data["sources"][self.version], strip_root=True, destination=self.source_folder)
```
Comment (SpaceIm) [inline recipes/glibmm/all/conanfile.py:93]:
```suggestion
        get(self, **self.conan_data["sources"][self.version], strip_root=True)
```

File: recipes/glibmm/all/conanfile.py [hunk 13]
```diff
@@ -43,129 +51,115 @@ def _giomm_lib(self):
...
 
             # when using cpp_std=c++NM the /permissive- flag is added which
             # attempts enforcing standard conformant c++ code
             # the problem is that older versions of Windows SDK is not standard
             # conformant! see:
             # https://developercommunity.visualstudio.com/t/error-c2760-in-combaseapih-with-windows-sdk-81-and/185399
-            files.replace_in_file(self, meson_build, "cpp_std=c++", "cpp_std=vc++")
+            replace_in_file(self, meson_build, "cpp_std=c++", "cpp_std=vc++")
 
     def configure(self):
```
Comment (SpaceIm) [inline recipes/glibmm/all/conanfile.py:135]: I wouldn't be against reorder a little bit methods by order of execution, it's super hard to follow global logic of this recipe with configur

File: recipes/glibmm/all/conanfile.py [hunk 14]
```diff
@@ -189,11 +184,11 @@ def package_info(self):
...
-            self.cpp_info.components[glibmm_component].requires = ["glib::gobject-2.0", "libsigcpp::sigc++-2.0"]
+            self.cpp_info.components[glibmm_component].requires = ["glib::gobject-2.0", "libsigcpp::libsigcpp"]
 
         self.cpp_info.components[giomm_component].set_property("pkg_config_name", giomm_component)
         self.cpp_info.components[giomm_component].libs = [giomm_component]
-        self.cpp_info.components[giomm_component].includedirs = [ os.path.join("include", giomm_component)]
+        self.cpp_info.components[giomm_component].includedirs = [os.path.join("include", giomm_component)]
         self.cpp_info.components[giomm_component].requires = [glibmm_component, "glib::gio-2.0"]
 
     def package_id(self):
```
Comment (SpaceIm) [inline recipes/glibmm/all/conanfile.py:194]: I can't review the line below, but there is a `if not self.options["glib"].shared` here to replace by `if not self.dependencies["glib"].options.shared`

File: recipes/glibmm/all/conanfile.py [hunk 15]
```diff
@@ -189,11 +184,11 @@ def package_info(self):
         if self._abi_version == "2.68":
             self.cpp_info.components[glibmm_component].requires = ["glib::gobject-2.0", "libsigcpp::sigc++"]
         else:
-            self.cpp_info.components[glibmm_component].requires = ["glib::gobject-2.0", "libsigcpp::sigc++-2.0"]
+            self.cpp_info.components[glibmm_component].requires = ["glib::gobject-2.0", "libsigcpp::libsigcpp"]
```
Comment (SpaceIm) [inline recipes/glibmm/all/conanfile.py:187]: If you look at libsigcpp recipe, this branching shouldn't exist, it should always be `self.cpp_info.components[glibmm_component].requires = ["glib::gobject-2.0", "libsigcpp::libsigcpp"]`. Existence of a component in libsigcpp is (or was) just a trick for target name of legacy cmake_find_package*.

File: recipes/glibmm/all/conanfile.py [hunk 16]
```diff
@@ -43,129 +51,115 @@ def _giomm_lib(self):
...
+    def generate(self):
+        deps = PkgConfigDeps(self)
+        deps.generate()
+
+        tc = MesonToolchain(self)
+        tc.project_options.update({
+            "build-examples": "false",
+            "build-documentation": "false",
+            "msvc14x-parallel-installable": "false",
+            "default_library": "shared" if self.options.shared else "static"
```
Comment (SpaceIm) [inline recipes/glibmm/all/conanfile.py:104]:
```suggestion
```

already managed by MesonToolchain

**Review by SpaceIm — APPROVED** [review]

**Review by uilianries — APPROVED** [review]
LGTM
